Error migrating Null collumns

When running scylla-migrator i am getting the following error on a table with 200 records that allows null values:
```
20/10/06 08:58:41 WARN TaskSetManager: Lost task 8.16 in stage 0.0 (TID 122, 172.17.0.3, executor 0): java.lang.RuntimeException: Error while encoding: java.lang.RuntimeException: org.apache.spark.unsafe.types.UTF8String is not a valid external type for schema of string
staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 0, app_id), StringType), true, false) AS app_id#0
staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 1, app_version_id), StringType), true, false) AS app_version_id#1
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 2, active), BooleanType) AS active#2
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else staticinvoke(class org.apache.spark.sql.catalyst.util.DateTimeUtils$, TimestampType, fromJavaTimestamp, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 3, created), TimestampType), true, false) AS created#3
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 4, destination), StringType), true, false) AS destination#4
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else mapobjects(MapObjects_loopValue0, MapObjects_loopIsNull0, ObjectType(class java.lang.Object), staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(lambdavariable(MapObjects_loopValue0, MapObjects_loopIsNull0, ObjectType(class java.lang.Object), true), StringType), true, false), validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 5, events), ArrayType(StringType,true)), None) AS events#5
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else staticinvoke(class org.apache.spark.sql.catalyst.util.DateTimeUtils$, TimestampType, fromJavaTimestamp, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 6, modified), TimestampType), true, false) AS modified#6
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 7, shared_secret_id), StringType), true, false) AS shared_secret_id#7
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:292)
	at org.apache.spark.sql.SparkSession$$anonfun$4.apply(SparkSession.scala:593)
	at org.apache.spark.sql.SparkSession$$anonfun$4.apply(SparkSession.scala:593)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at com.datastax.spark.connector.util.CountingIterator.next(CountingIterator.scala:16)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.next(GroupingBatchBuilder.scala:105)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.next(GroupingBatchBuilder.scala:30)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.foreach(GroupingBatchBuilder.scala:30)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$2.apply(TableWriter.scala:241)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$2.apply(TableWriter.scala:210)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:112)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:111)
	at com.datastax.spark.connector.cql.CassandraConnector.closeResourceAfterUse(CassandraConnector.scala:129)
	at com.datastax.spark.connector.cql.CassandraConnector.withSessionDo(CassandraConnector.scala:111)
	at com.datastax.spark.connector.writer.TableWriter.writeInternal(TableWriter.scala:210)
	at com.datastax.spark.connector.writer.TableWriter.insert(TableWriter.scala:188)
	at com.datastax.spark.connector.writer.TableWriter.write(TableWriter.scala:175)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:38)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:38)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: org.apache.spark.unsafe.types.UTF8String is not a valid external type for schema of string
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.MapObjects_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.writeFields_0_1$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:289)
	... 30 more
```

@iravid This is our setup that causes the problem.


# table definition
```
CREATE TABLE qa2.exclusions (
    sku text,
    exclusion_type text,
    exclusions list<text>,
    PRIMARY KEY (sku, exclusion_type)
) WITH CLUSTERING ORDER BY (exclusion_type ASC)
    AND bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.DeflateCompressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
```
# exception trace
```
java.lang.RuntimeException: Error while encoding: java.lang.RuntimeException: org.apache.spark.unsafe.types.UTF8String is not a valid external type for schema of string
staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 0, sku), StringType), true, false) AS sku#0
staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 1, exclusion_type), StringType), true, false) AS exclusion_type#1
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else mapobjects(MapObjects_loopValue0, MapObjects_loopIsNull0, ObjectType(class java.lang.Object), staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(lambdavariable(MapObjects_loopValue0, MapObjects_loopIsNull0, ObjectType(class java.lang.Object), true), StringType), true, false), validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 2, exclusions), ArrayType(StringType,true)), None) AS exclusions#2
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:292)
	at org.apache.spark.sql.SparkSession$$anonfun$4.apply(SparkSession.scala:594)
	at org.apache.spark.sql.SparkSession$$anonfun$4.apply(SparkSession.scala:594)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at com.datastax.spark.connector.util.CountingIterator.next(CountingIterator.scala:16)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.next(GroupingBatchBuilder.scala:105)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.next(GroupingBatchBuilder.scala:30)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.foreach(GroupingBatchBuilder.scala:30)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$2.apply(TableWriter.scala:241)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$2.apply(TableWriter.scala:210)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:112)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:111)
	at com.datastax.spark.connector.cql.CassandraConnector.closeResourceAfterUse(CassandraConnector.scala:129)
	at com.datastax.spark.connector.cql.CassandraConnector.withSessionDo(CassandraConnector.scala:111)
	at com.datastax.spark.connector.writer.TableWriter.writeInternal(TableWriter.scala:210)
	at com.datastax.spark.connector.writer.TableWriter.insert(TableWriter.scala:188)
	at com.datastax.spark.connector.writer.TableWriter.write(TableWriter.scala:175)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:38)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:38)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: org.apache.spark.unsafe.types.UTF8String is not a valid external type for schema of string
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.MapObjects_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:289)
	... 30 more
```

--- a/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
+++ b/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
@@ -222,16 +222,26 @@ object Cassandra {
       .select(selection.columnRefs: _*)
       .asInstanceOf[RDD[Row]]
       .map { row =>
-        Row.fromSeq(
-          row.toSeq.map {
-            // We're using a low-level API of the Cassandra connector which produces UTF8String
-            // objects for the various textual data types in Cassandra. Spark forbids this type
-            // from showing up in DataFrames (it usually converts it internally), so we need to
-            // convert it to a String ourselves.
-            case x: UTF8String => x.toString
-            case x             => x
-          }
-        )
+        // We're using a low-level API of the Cassandra connector which produces UTF8String
+        // objects for the various textual data types in Cassandra. Spark forbids this type
+        // from showing up in DataFrames (it usually converts it internally), so we need to
+        // convert it to a String ourselves. Unfortunately we need to also recurse through the
+        // composite data types that can be produced from Cassandra, so this logic is slightly
+        // more involved than simply mapping.
+        lazy val convertUTF8String: Any => AnyRef = {
+          case x: UTF8String => x.toString
+          case set: Set[_]   => set.map(convertUTF8String)
+          case list: List[_] => list.map(convertUTF8String)
+          case map: Map[_, _] =>
+            map.map {
+              case (k, v) => convertUTF8String(k) -> convertUTF8String(v)
+            }
+          case UDTValue(names, values) => UDTValue(names, values map convertUTF8String)
+          case tuple: TupleValue       => TupleValue(tuple.values.map(convertUTF8String): _*)
+          case x                       => x.asInstanceOf[AnyRef]
+        }
+
+        Row.fromSeq(row.toSeq.map(convertUTF8String))
       }
 
     val resultingDataframe = adjustDataframeForTimestampPreservation(

--- a/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
+++ b/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
@@ -233,21 +233,21 @@ object Cassandra {
         // done in the SourceRelation connector of the Dataframe API. So we have to replicate
         // them here. Future versions of the migrator will use the DataFrame API directly to
         // avoid this replication.
-        lazy val convertUTF8String: Any => AnyRef = {
+        lazy val convertRowTypes: Any => AnyRef = {
           case x: UTF8String => x.toString
-          case set: Set[_]   => set.map(convertUTF8String)
-          case list: List[_] => list.map(convertUTF8String)
+          case set: Set[_]   => set.map(convertRowTypes)
+          case list: List[_] => list.map(convertRowTypes)
           case map: Map[_, _] =>
             map.map {
-              case (k, v) => convertUTF8String(k) -> convertUTF8String(v)
+              case (k, v) => convertRowTypes(k) -> convertRowTypes(v)
             }
-          case ab: ArrayBuffer[_] => ab.map(convertUTF8String)
-          case udt: UDTValue      => Row.fromSeq(udt.columnValues.map(convertUTF8String))
-          case tuple: TupleValue  => Row.fromSeq(tuple.values.map(convertUTF8String))
+          case ab: ArrayBuffer[_] => ab.map(convertRowTypes)
+          case udt: UDTValue      => Row.fromSeq(udt.columnValues.map(convertRowTypes))
+          case tuple: TupleValue  => Row.fromSeq(tuple.values.map(convertRowTypes))
           case x                  => x.asInstanceOf[AnyRef]
         }
 
-        Row.fromSeq(row.toSeq.map(convertUTF8String))
+        Row.fromSeq(row.toSeq.map(convertRowTypes))
       }
 
     val resultingDataframe = adjustDataframeForTimestampPreservation(

--- a/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
+++ b/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
@@ -15,6 +15,8 @@ import org.apache.spark.sql.types.{ LongType, StructField, StructType }
 import org.apache.spark.sql.{ DataFrame, Row, SparkSession }
 import org.apache.spark.unsafe.types.UTF8String
 
+import scala.collection.mutable.ArrayBuffer
+
 object Cassandra {
   val log = LogManager.getLogger("com.scylladb.migrator.readers.Cassandra")
 
@@ -222,12 +224,15 @@ object Cassandra {
       .select(selection.columnRefs: _*)
       .asInstanceOf[RDD[Row]]
       .map { row =>
-        // We're using a low-level API of the Cassandra connector which produces UTF8String
-        // objects for the various textual data types in Cassandra. Spark forbids this type
-        // from showing up in DataFrames (it usually converts it internally), so we need to
-        // convert it to a String ourselves. Unfortunately we need to also recurse through the
-        // composite data types that can be produced from Cassandra, so this logic is slightly
-        // more involved than simply mapping.
+        // We need to handle three conversions here that are not done for us:
+        // - UTF8Strings to plain Strings
+        // - UDTValue to Row
+        // - TupleValue to Row
+        //
+        // We're using the RDD API of the Cassandra connector, and these conversions are
+        // done in the SourceRelation connector of the Dataframe API. So we have to replicate
+        // them here. Future versions of the migrator will use the DataFrame API directly to
+        // avoid this replication.
         lazy val convertUTF8String: Any => AnyRef = {
           case x: UTF8String => x.toString
           case set: Set[_]   => set.map(convertUTF8String)
@@ -236,9 +241,10 @@ object Cassandra {
             map.map {
               case (k, v) => convertUTF8String(k) -> convertUTF8String(v)
             }
-          case UDTValue(names, values) => UDTValue(names, values map convertUTF8String)
-          case tuple: TupleValue       => TupleValue(tuple.values.map(convertUTF8String): _*)
-          case x                       => x.asInstanceOf[AnyRef]
+          case ab: ArrayBuffer[_] => ab.map(convertUTF8String)
+          case udt: UDTValue      => Row.fromSeq(udt.columnValues.map(convertUTF8String))
+          case tuple: TupleValue  => Row.fromSeq(tuple.values.map(convertUTF8String))
+          case x                  => x.asInstanceOf[AnyRef]
         }
 
         Row.fromSeq(row.toSeq.map(convertUTF8String))

